eip-0008.md

```diff
@@ -0,0 +1,479 @@
+      BTW should be adjustable via voting). The time budget for block assembly and
+      validation itself is the fraction of this time (say 30 seconds), as the block must
+      also be distributed across the network and validated by all nodes. This is necessary
+      to avoid network splits.
+      - The maximum number of transactions in a block is limited by the `MaxBlockCost`
+      network parameter. This parameter can be increased by miners via voting (practically
+      indefinitely), as long as the resulting blocks don't exceed the validation time
+      limit.
+
+Under these limits, miners as a community, are interested to improve throughput of block
+assembly and validation software to process more transactions per second, with higher
+total cost within the given time limits. The proposed JIT costing protocol for extended
+UTXO model of Ergo enables a whole range of potential optimizations. This will be
+especially important when the emission is over and the transaction fees along with storage
+rent will be the only reward for miners.
+
+### JIT costing of ErgoTree
+
+Ethereum contracts are compiled to bytecode instructions, which are then executed by
+Ethereum VM as part of a transaction execution. Each instruction has an associated cost (aka
+'gas') and the cost of the whole transaction is accrued as instructions are executed one
+by one until either transaction finishes of gasLimit is reached, in which case the
+transaction is aborted and the `gasLimit * gasPrice` number of Wei is transferred to the
+miner.
+
+In contrast, Ergo doesn't use VM to execute contracts which protect boxes. The contract in
+Ergo is stored as serialized _typed abstract syntax tree_ (ErgoTree). The bytes of the
+contract are deserialized into ErgoTree data structure. 
+
+ErgoTree is designed for direct and efficient execution by the
+interpreter using a Context data structure to access the blockchain data. The interpreter
+doesn't have direct access to the blockchain, and rely only on the Context to execute ErgoTree.
+The original interpreter of the language is extended to
+accumulate the execution cost during evaluation of ErgoTree. Thus when the interpreter 
+finishes the execution of ErgoTree, the result is a pair `(R,C)`, where `R` is true/false
+boolean value and `C` is the cost accumulated during execution of `R`. If however during
+the execution of ErgoTree the accumulated cost exceeds at some point the `costLimit`
+value, then exception is thrown, which is interpreted as over-expensive script.
+
+An algorithm for execution of ErgoTree with JIT cost estimation is described in [ErgoTree
+Specification](https://ergoplatform.org/docs/ergotree.pdf).
+
+### Initial tests of transaction validity
+
+This conditions should be checked for every well-formed transaction.
+
+- All checks performed in Protocol v1
+- INPUTS contain at least one `p2pk(senderPk)` box
+- current height is within `feeOut.R6` interval
```

Comment:
Please note that the fee box is not a part of the protocol.

Reply:
This part of JIT costing depends on the concrete implementation of miner fees. 
They are both "a part" or "not a part" of the protocol  (whichever the definition of the protocol).
If some other fee protocol will be developed in future (if any), it should keep JIT costing in mind. 
I consider JIT costing is more important than some abstract fee protocol flexibility. 
Even if the new version of fees protocol will be added, it can always be made compatible with the needs of JIT costing.

Reply:
"All checks performed in Protocol v1" means protocol changes, right? On the definition of the protocol, see V() predicate in GKL 15.

Reply:
No, it means currently active protocol. 
By this item I refer to all existing validation checks collectively. 
Other items are new.
Clarified this in the text.